solarforecastarbiter/validation/tasks.py

```diff
@@ -652,9 +580,144 @@ def apply_validation(observation, observation_values):
+    # split data to post at most one week at a time
+    # ~10,000 pts of 1min data
+    week_int = (gid.index.week + gid.index.year).values
+    # combine the continuous groups with groups of weeks
+    # gid is unique for each group since week_int and cumsum
+    # increase monotonically and are positive
+    gid += week_int
+    observation_values['gid'] = gid
+    for _, group in observation_values.groupby('gid'):
+        session.post_observation_values(observation.observation_id,
+                                        group[['value', 'quality_flag']],
+                                        params='donotvalidate')
+
+
+def _validate_post(session, observation, start, end):
+    logger.info('Validating data for %s from %s to %s',
+                observation.name, start, end)
+    observation_values = session.get_observation_values(
+        observation.observation_id, start, end)
+    validated = apply_validation(observation, observation_values)
+    return _group_continuous_week_post(
+        session, observation, validated)
+
+
+def _find_unvalidated_time_ranges(session, observation, min_start, max_end):
+    """Find the time ranges where the observation data needs to have
+    daily validation applied. Extend to next day midnight so daily
+    validation can be applied even since it requires >= 1 day of data
+    """
+    tz = observation.site.timezone
+    dates = np.array(sorted(
+        session.get_observation_values_not_flagged(
+            observation_id=observation.observation_id,
+            start=min_start,
+            end=max_end,
+            flag=(
+                quality_mapping.DAILY_VALIDATION_FLAG |
+                quality_mapping.LATEST_VERSION_FLAG
+            ),
+            timezone=tz)
+        ))
+    if len(dates) == 0:
+        return
+
+    breaks = np.diff(dates).astype('timedelta64[D]') - 1
+
+    def fl(prev, ind):
+        first = pd.Timestamp(dates[prev]).tz_localize(tz)
+        last = (pd.Timestamp(dates[ind])
+                .tz_localize(tz) + pd.Timedelta('1D'))
+        return first, last
+
+    prev = 0
+    for ind in np.nonzero(breaks)[0]:
+        first, last = fl(prev, ind)
+        yield first, last
+        prev = ind + 1
+
+    first, last = fl(prev, -1)
+    yield first, last
+
+
+def _split_validation(session, observation, start, end, only_missing):
+    if not only_missing:
+        return _validate_post(session, observation, start, end)
+
+    for _start, _end in _find_unvalidated_time_ranges(
+            session, observation, start, end):
+        _validate_post(session, observation, _start, _end)
+
+
+def fetch_and_validate_observation(access_token, observation_id, start, end,
+                                   only_missing=False, base_url=None):
+    """Task that will run immediately after Observation values are
+    uploaded to the API to validate the data. If over a day of data is
+    present, daily validation will be applied.
```

Comment:
document behavior for last day of a multiday series that has only a partial day.